### PR TITLE
Select to poll

### PR DESCRIPTION
**Pull request by igron99**
fd_set can not contain fd >= 1024.
Fixed with poll

### CODE REVIEW

**Comment by igron99** [comment]
```cannot find library libmsgpuck.so.0 needed by debian/libtarantool-c/usr/lib/libtarantool.so.2.0.0```

File: tnt/tnt_io.c [hunk 1]
```diff
@@ -138,15 +139,13 @@ tnt_io_connect_do(struct tnt_stream_net *s, struct sockaddr *addr,
...
 					}
 					/* calculate timeout last time */
-					int64_t passd_usec = (curr.tv_sec - start_connect.tv_sec) * micro +
-						(curr.tv_usec - start_connect.tv_usec);
-					int64_t curr_tmeout = passd_usec - tmout_usec;
-					if (curr_tmeout <= 0) {
+					int passd_msec = (curr.tv_sec - start_connect.tv_sec) * milli +
+						(curr.tv_usec - start_connect.tv_usec) / milli;
+					int tmout = passd_msec - tmout_msec;
+					if (tmout <= 0) {
```
[the remark this answers is not in the document]
Comment (igron99) [reply tnt/tnt_io.c]: fixed

File: tnt/tnt_io.c [hunk 2]
```diff
@@ -138,15 +139,13 @@ tnt_io_connect_do(struct tnt_stream_net *s, struct sockaddr *addr,
...
-					if (curr_tmeout <= 0) {
+					int passd_msec = (curr.tv_sec - start_connect.tv_sec) * milli +
+						(curr.tv_usec - start_connect.tv_usec) / milli;
+					if (passd_msec >= tmout_msec) {
 						/* timeout */
 						return TNT_ETMOUT;
-					}
-					tmout.tv_sec = curr_tmeout / micro;
-					tmout.tv_usec = curr_tmeout % micro;
+                                        }
```
Comment (Totktonada) [inline tnt/tnt_io.c]: Nit: spaces instead of tabs.

File: tnt/tnt_io.c [hunk 3]
```diff
@@ -138,15 +139,13 @@ tnt_io_connect_do(struct tnt_stream_net *s, struct sockaddr *addr,
 						return TNT_ESYSTEM;
 					}
 					/* calculate timeout last time */
-					int64_t passd_usec = (curr.tv_sec - start_connect.tv_sec) * micro +
-						(curr.tv_usec - start_connect.tv_usec);
-					int64_t curr_tmeout = passd_usec - tmout_usec;
-					if (curr_tmeout <= 0) {
+					int passd_msec = (curr.tv_sec - start_connect.tv_sec) * milli +
```
Comment (Totktonada) [inline tnt/tnt_io.c]: Nit: >80 symbols.

Comment (igron99) [reply tnt/tnt_io.c]: The line, that i replaced was > 80 but < 120 simbols too. Please suggest other way.

Comment (Totktonada) [reply tnt/tnt_io.c]: Our C code style suggests to limit line length to 80 symbols: [1]. If the previous code was not formatted properly, so it is bad, but is not the reason to break code style in the new code.

[1]: https://tarantool.io/en/doc/1.9/dev_guide/c_style_guide/#chapter-2-breaking-long-lines-and-strings

File: tnt/tnt_io.c [hunk 4]
```diff
@@ -113,22 +114,22 @@ tnt_io_connect_do(struct tnt_stream_net *s, struct sockaddr *addr,
...
 		/* get start connect time */
 		struct timeval start_connect;
 		if (gettimeofday(&start_connect, NULL) == -1) {
 			s->errno_ = errno;
 			return TNT_ESYSTEM;
 		}
 		/* set initial timer */
-		struct timeval tmout;
-		memcpy(&tmout, &s->opt.tmout_connect, sizeof(tmout));
+		int tmout = s->opt.tmout_connect.tv_sec * milli +
```
Comment (Totktonada) [inline tnt/tnt_io.c]: Nit: trailing whitespace.

File: tnt/tnt_io.c (hunk 3 above)
Comment (igron99) [reply tnt/tnt_io.c]: i mean, i do not know how to make line < 80 symbols gracefully

Comment (Totktonada) [reply tnt/tnt_io.c]: It likely will not be beautiful. Ideally we should decrease indent level by extracting code pieces into functions. But this is a kind of refactoring and usually is not mandatory (or even discouraged if is massive) within a feature/bug scope.

Don't worry much about that, this is the small piece of the code.

Comment (igron99) [reply tnt/tnt_io.c]: Thank you
Fixed

**Comment by Totktonada** [comment]
Use macroses to prettify the code and squashed commits into the one on `gh-102-replace-select-with-poll` branch. Commit: b4c6704b4a6bff1e43971f6813774ce2079347a8.

@bigbes Can you, please, look into that?

Don't sure how to write a test, btw, so leave it now w/o the test case. I tried to check it like so (based on readme example):

```
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <sys/time.h>
#include <errno.h>

#include <tarantool/tarantool.h>
#include <tarantool/tnt_net.h>
#include <tarantool/tnt_opt.h>

int main()
{
	const char *uri = "localhost:3301";
	struct tnt_stream *tnt = tnt_net(NULL);
	assert(tnt != NULL);
	assert(tnt_set(tnt, TNT_OPT_URI, uri) == 0);
	assert(tnt_set(tnt, TNT_OPT_SEND_BUF, 0) == 0);
	assert(tnt_set(tnt, TNT_OPT_RECV_BUF, 0) == 0);
	struct timeval tv;
	tv.tv_sec = 10;
	tv.tv_usec = 0;
	assert(tnt_set(tnt, TNT_OPT_TMOUT_CONNECT, &tv) == 0);
	int ret;
	if ((ret = tnt_connect(tnt)) != 0) {
		perror("tnt_connect");
		assert(0);
	}
	assert(tnt_ping(tnt) == 0);
	struct tnt_reply *reply = tnt_reply_init(NULL);
	assert(tnt->read_reply(tnt, reply) == 0);
	tnt_reply_free(reply);
	tnt_close(tnt);
	tnt_stream_free(tnt);
	return 0;
}
```

I down the tarantool on localhost:3301 and run the test above. `connect` says EINPROGRESS, `poll` returns with 1, `getsockopt(..., SO_ERROR, ...)` says ECONNREFUSED. Is it expected behaviour? Maybe we should move the `getsockopt` check under the loop?

**Comment by Totktonada** [comment]
Eugine (@bigbes) commented that the connect timeout intentionally reflects socket API and only set timeout to wait SYN-ACK. When RST is got instead we have to return the error immediately.

I'm going to write integration tests to check timeout and >1023 connections.

**Comment by Totktonada** [comment]
Superseded with PR #109.

**Pull request closed without merging**